### PR TITLE
Builder tests

### PR DESCRIPTION
**Pull request by hemanik**
Includes tests for Builder API.

### CODE REVIEW

File: arquillian-core-reporter/arquillian-core-reporter-impl/pom.xml [hunk 1]
```diff
@@ -53,8 +53,38 @@
...
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (MatousJobanek) [inline arquillian-core-reporter/arquillian-core-reporter-impl/pom.xml]: I think that this shouldn't be part of this PR as it is only about BuilderTests which are  in reporter-core

File: core/reporter-api/src/main/java/org/arquillian/reporter/api/model/entry/KeyValueEntry.java [hunk 2]
```diff
@@ -57,6 +57,7 @@ public KeyValueEntry(String key, String value) {
         this.value = new StringEntry(value);
     }
 
+
```
Comment (MatousJobanek) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/entry/KeyValueEntry.java]: Please, clean the PR so it doesn't contain any formatting changes in files that haven't been changed

File: core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
+import org.arquillian.reporter.api.builder.report.ReportBuilder;
 import org.arquillian.reporter.api.model.StringKey;
 import org.arquillian.reporter.api.model.UnknownStringKey;
 import org.arquillian.reporter.api.model.entry.Entry;
-import org.arquillian.reporter.api.builder.report.ReportBuilder;
 import org.arquillian.reporter.api.utils.Validate;
 
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (MatousJobanek) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java]: Please, clean the PR so it doesn't contain any formatting changes in files that haven't been changed

File: core/reporter-impl/pom.xml [hunk 4]
```diff
@@ -78,6 +78,12 @@
             <artifactId>mockito-core</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>21.0</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (MatousJobanek) [inline core/reporter-impl/pom.xml]: This is because of the line: https://github.com/arquillian/arquillian-reporter/pull/15/files#diff-1f43dc9359d1729a792ef0c4c821e8c5R74 right?
I think that it doesn't bring any additional value, so we can use just normal HashMap and to not add an additional dependency.
Correct me If I'm wrong

File: core/reporter-impl/src/test/java/org/arquillian/reporter/impl/model/report/AllReportsMergeTest.java [hunk 5]
```diff
@@ -17,6 +13,10 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
```
Comment (MatousJobanek) [inline core/reporter-impl/src/test/java/org/arquillian/reporter/impl/model/report/AllReportsMergeTest.java]: Please, clean the PR so it doesn't contain any formatting changes in files that haven't been changed

File: core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java (hunk 3 above)
Comment (bartoszmajsak) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java]: I would suggest to share `.editorconfig` file in the project. We have it in few places already.

File: core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                .hasSubReportsContaining(Arrays.asList(new BasicReport("sub report 1"), new BasicReport("sub report 2")))
+                .hasNumberOfEntries(16)
+                .hasNumberOfSubReports(2)
+                .isInstanceOf(reportClass);
+    }
+
+    @Override
+    protected void addAdditionalExtensions(List<Class<?>> extensions) {
+
+    }
```
Comment (bartoszmajsak) [inline core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java]: Why do we need those empty methods? Maybe instead of inheritance we can try with composition and have a class which implements this functionality instead? With the current shape it looks a bit odd.

File: core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java (hunk 3 above)
Comment (MatousJobanek) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java]: Thanks for your suggestion - I'll do

File: core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java (hunk 6 above)
Comment (MatousJobanek) [inline core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java]: inheritance - because we need to do some preparation/registration in before methods - the AbstractReporterTestBase extends AbstractTestTestBase from core and adds some additional functionality.
Or, I could change the method not to be abstract and leave it empty in the base class with an opportunity of overriding it.

Comment (MatousJobanek) [inline core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java]: But I do agree that it looks odd. My intention was to make it clearly visible that you can register additional implementation so you don't have to go to the base class and find for the right methods...

File: core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java (hunk 3 above)
Comment (MatousJobanek) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java]: Correct me if I'm wrong, but .editorconfig doesn't provide a formatting for an order of imports, does it?

Comment (bartoszmajsak) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java]: Yeah, if we want to go deeper to IDE settings we can share xmls... but maybe it's not really needed

File: core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java (hunk 6 above)
Comment (bartoszmajsak) [inline core/reporter-impl/src/test/java/org/arquillian/reporter/impl/builder/AllBuildersTest.java]: > inheritance - because we need to do some preparation/registration in before methods - the AbstractReporterTestBase extends AbstractTestTestBase from core and adds some additional functionality.

Or we can have a class which has this logic and just call it in before method of the abstract class instead? Another way of delegating responsibilities

File: core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java (hunk 3 above)
Comment (bartoszmajsak) [inline core/reporter-api/src/main/java/org/arquillian/reporter/api/model/report/AbstractReport.java]: That's true :) I thought more of

For imports as well as curly braces and so on we have to look for some other approach. maybe sharing IDE specific settings will be enough
